languageserver/src/main/scala/langserver/core/Connection.scala

```diff
@@ -16,23 +14,25 @@ import langserver.messages._
 import langserver.types._
 import play.api.libs.json._
 import com.dhpcs.jsonrpc.JsonRpcMessage._
+import monix.eval.Task
+import monix.execution.Scheduler
 
 /**
  * A connection that reads and writes Language Server Protocol messages.
  *
+ * @param s thread pool to run tasks with.
  * @note Commands are executed asynchronously via a thread pool
  * @note Notifications are executed synchronously on the calling thread
  * @note The command handler returns Any because sometimes response objects can't be part
  *       of a sealed hierarchy. For instance, goto definition returns a {{{Seq[Location]}}}
  *       and that can't subclass anything other than Any
  */
-class Connection(inStream: InputStream, outStream: OutputStream)(val commandHandler: (String, ServerCommand) => Any)
+abstract class Connection(inStream: InputStream, outStream: OutputStream)(implicit s: Scheduler)
```

Comment:
How about making this a `trait`? TBH I never understood the `abstract class` vs `trait` thing in Scala, hence I always used traits.

Reply:
I prefer abstract class unless I expect it to be mixed into another class. abstract classes can accept parameters (traits get that ability too in dotty) and they also make it easier to keep compatibility pre-2.12 (although that's irrelevant here). I think it's best to keep this as a class here, I can make it non-abstract by implementing commandHandler, but it doesn't really matter for us much here.